### PR TITLE
Verify binaries using hashes

### PR DESCRIPTION
**Pull request by mosuem**
When downloading pre-compiled binaries from the CDN, check the hashes of the binaries to verify the correct file was downloaded.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: pkgs/intl4x/build.dart [hunk 1]
```diff
@@ -73,25 +82,46 @@ sealed class BuildMode {
...
+
+    await writeFile(unzippedibrary, file);
+    final fileHash =
+        sha256.convert(unzippedibrary.readAsBytesSync()).toString();
+    if (fileHash == fileHashes[config.target]) {
+      final library = File(libPath);
+      library.parent.createSync(recursive: true);
+      unzippedibrary.copySync(libPath);
+    } else {
+      throw ArgumentError('The pre-built binary for the target '
```
Comment (sigurdm) [inline pkgs/intl4x/build.dart:114]: Drive-by nit.

Perhaps this should be an exception (or just exit(-1)) as it is not a programmatic error, but an active check.

File: pkgs/intl4x/tool/generate_hashes.dart [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  await dynamicLibrary.create();
+  final httpClient = HttpClient();
+  for (final os in OS.values) {
+    for (final architecture in Architecture.values) {
+      final target = '${os}_$architecture';
+      print('Checking hash for $target');
+      final uri = Uri.parse(
+          'https://github.com/dart-lang/i18n/releases/download/$version/$target');
+      final request = await httpClient.getUrl(uri);
+      final response = await request.close();
```
Comment (robertbastian) [inline pkgs/intl4x/tool/generate_hashes.dart:21]: Why is this generating hashes for downloaded artifacts? You might already be mitm'ed, it's better to calculate hashes when building the binaries.

Comment (mosuem) [reply pkgs/intl4x/tool/generate_hashes.dart:21]: To compare the hash with the stored one - this should guard against mitm if I am not mistaken.

Comment (mosuem) [reply pkgs/intl4x/tool/generate_hashes.dart:21]: Sorry, I just saw this is in the generating hashes script. This is run only for updates, so it should be fine. Could you cross-compile all targets on a single machine to get the hashes?

Comment (robertbastian) [reply pkgs/intl4x/tool/generate_hashes.dart:21]: You're checking in the result of this, right?

Comment (mosuem) [reply pkgs/intl4x/tool/generate_hashes.dart:21]: Yes.

Comment (robertbastian) [reply pkgs/intl4x/tool/generate_hashes.dart:21]: You can pass artifacts between GH actions jobs, so there could be a job at the end that calculates the hashes from the actual binaries before they touch the network.

Comment (mosuem) [reply pkgs/intl4x/tool/generate_hashes.dart:21]: Yeah a test could be added to the Github action.

**Review by robertbastian — APPROVED** [review]
